### PR TITLE
global-plugins: improvements, subpositions, fixes for cachefilter

### LINKED ISSUE #1241: global plugins: invalid read
@markus2330 you mentioned a memleak in #874.

Did you mean this: https://build.libelektra.org/job/elektra-mergerequests-stable/811/artifact/build/Testing/20161227-2311/DynamicAnalysis.xml/*view*/

### PR DESCRIPTION
**Pull request by mpranj**
# Purpose

Improvements for global-plugins. The purpose is to get them at least ready to support cachefilter.

Fixes #1241.

# Checklist

- [x] commit messages are fine (with references to issues)
- [x] I ran all tests and everything went fine

### CODE REVIEW

**Comment by markus2330** [comment]
Thank you, looks good! elektra-mergerequests and elektra-gcc-configure-mingw-w64 #1254 are broken on master.

**Comment by mpranj** [comment]
Thanks, I was wondering what's going on.

@Namoshek can you check if cachefilter would work on 145d6faf72e5033504d6ad785d7bba8348be6006 and/or specify what's not working so I can fix it?

**Comment by markus2330** [comment]
@mpranj did you already tested whats discussed in #1072? Please also automate the tests there.

**Comment by mpranj** [comment]
@markus2330 I get the same result as you in #1072.

Activating cachefilter as a default global-plugin and running all tests, it does seem that we need to fix a few things:
```
[...]
131/132 Test #131: testkdb_nested .....................***Failed    0.34 sec
Running main() from gtest_main.cc
[==========] Running 6 tests from 1 test case.
[----------] Global test environment set-up.
[----------] 6 tests from Nested
[ RUN      ] Nested.GetSetNothing
[       OK ] Nested.GetSetNothing (56 ms)
[ RUN      ] Nested.GetSetRoot
[       OK ] Nested.GetSetRoot (55 ms)
[ RUN      ] Nested.GetSetBelow
/Users/mpranj/workspace/libelektra/tests/kdb/testkdb_nested.cpp:117: Failure
Value of: 715
Expected: ks.size ()
Which is: 1
did not keep key at getsystem/tests/kdb/below/key

[  FAILED  ] Nested.GetSetBelow (53 ms)
[ RUN      ] Nested.RemoveFiles
[...]
```

Surprisingly many tests run fine though.

**Comment by mpranj** [comment]
As discussed in the last meeting it's not a complete implementation of all global plugins postitions. It fixes some memory problems and should be able to support cachefilter now.

@KurtMi please review.

File: src/libs/elektra/kdb.c [hunk 1]
```diff
@@ -865,6 +883,8 @@ int kdbGet (KDB * handle, KeySet * ks, Key * parentKey)
 		splitMerge (split, ks);
 	}
 
+	elektraGlobalGet (handle, ks, parentKey, POSTGETSTORAGE, MAXONCE);
```
Comment (KurtMi) [inline src/libs/elektra/kdb.c]: Is there not a `INIT` and `DEINIT` missing?

File: src/libs/elektra/kdb.c [hunk 2]
```diff
@@ -1163,7 +1179,11 @@ int kdbSet (KDB * handle, KeySet * ks, Key * parentKey)
 	int errnosave = errno;
 	Key * initialParent = keyDup (parentKey);
 
-	ELEKTRA_LOG ("now in new kdbSet (%s)\n", keyName (parentKey));
+	ELEKTRA_LOG ("now in new kdbSet (%s) %p %zd", keyName (parentKey), (void *)handle, ksGetSize (ks));
+
+	elektraGlobalSet (handle, ks, parentKey, PRESETSTORAGE, MAXONCE);
```
Comment (KurtMi) [inline src/libs/elektra/kdb.c]: Is there not a `INIT` and `DEINIT` missing?

File: src/libs/elektra/kdb.c [hunk 3]
```diff
@@ -1222,20 +1242,24 @@ int kdbSet (KDB * handle, KeySet * ks, Key * parentKey)
 		copyError (parentKey, oldError);
```
Comment (KurtMi) [inline src/libs/elektra/kdb.c]: can not comment in line 1228 so i put it here:
There is a return in 1228, should there not be before the return call some hook calls (`postgetstorage`)?

File: src/libs/elektra/kdb.c [hunk 4]
```diff
@@ -1163,7 +1179,11 @@ int kdbSet (KDB * handle, KeySet * ks, Key * parentKey)
 	int errnosave = errno;
 	Key * initialParent = keyDup (parentKey);
 
-	ELEKTRA_LOG ("now in new kdbSet (%s)\n", keyName (parentKey));
+	ELEKTRA_LOG ("now in new kdbSet (%s) %p %zd", keyName (parentKey), (void *)handle, ksGetSize (ks));
```
Comment (KurtMi) [inline src/libs/elektra/kdb.c]: Generally: is the `SETSTORAGE` position not missing?

File: src/libs/elektra/kdb.c (hunk 3 above)
Comment (mpranj) [inline src/libs/elektra/kdb.c]: We have `presetstorage` which is done before (after the sanity checks like handle not NULL).

Other than that, I don't know of a hook position which suits that place.

**Comment by mpranj** [comment]
@KurtMi thanks for the review. The INIT/DEINIT was indeed missing.

In general this PR should just fix memory handling issues and get it ready for `cachefilter`. The other positions will be added later. Adding positions is really less of a problem than knowing where we really need a hook position.

My goal is to use this as a base to build the `mmap-cache` plugin. While developing it I may need to make some changes to the hooks anyway. I don't think it makes sense to finalize the hooks API without having one plugin utilizing it properly.

**Comment by markus2330** [comment]
> I don't think it makes sense to finalize the hooks API without having one plugin utilizing it properly.

Ack but for the currently defined hooks we already have plugins?

@KurtMi Did you also try to use semlock?

**Comment by KurtMi** [comment]
> @KurtMi Did you also try to use semlock?

Yes, but it did not work, I think there are some placements missing, I tried to fix it but failed, we should try in a later version of the global plugins.

Btw: is there a syntax for the placements in the `README.md`, something like:
```
- infos/placements = presetstorage/init postcommit/deinit postrollback/deinit getstorage/init getstorage/deinit
```

**Comment by mpranj** [comment]
@KurtMi haven't checked whether it works, but that should be the syntax when it's done, yes. Thanks for the review.

**Comment by markus2330** [comment]
@KurtMi can you elaborate what exactly did not work? (Maybe create an issue?)

**Comment by markus2330** [comment]
@mpranj Thank you, finally it is done!

Hoping to see a new PR with further improvements soon!

I removed the cachefilter branch now.

**Comment by mpranj** [comment]
@markus2330 thanks!

**Pull request merged**